src/ray/gcs/gcs_server/gcs_kv_manager.cc

```diff
@@ -33,6 +33,25 @@ void GcsInternalKVManager::HandleInternalKVGet(
       }));
 }
 
+void GcsInternalKVManager::HandleInternalKVMultiGet(
+    const rpc::InternalKVMultiGetRequest &request, rpc::InternalKVMultiGetReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  auto cnt = std::make_shared<int>(request.keys_size());
+  for (int i = 0; i < request.keys_size(); ++i) {
+    std::vector<std::string> cmd = {"HGET", request.keys(i), "value"};
+    RAY_CHECK_OK(redis_client_->GetPrimaryContext()->RunArgvAsync(
+        cmd, [i, reply, send_reply_callback, cnt](auto redis_reply) {
+          if (!redis_reply->IsNil()) {
```

Comment:
Is this callback running in the main io service thread?

Reply:
yes and that's bad due to our model

Reply:
The callback inside doesn't touch any private field right? Can you add comment to warn not to do that?

Reply:
sure. btw, i think in either way it won't be fixed easily (in place running or having a separate thread running this)